### PR TITLE
add node-module option for node.__file/dirname

### LINKED ISSUE #14072: ESM builds should substitute `__filename` and `__dirname` references in CommonJS
## Feature request

**What is the expected behavior?**

When building an ESM build in Webpack, free `__filename` and `__dirname` references in CommonJS files should be supported.

**What is motivation or use case for adding/changing the behavior?**

See eg https://github.com/vercel/ncc/issues/749 for an ncc bug here.

**How should this be implemented in your opinion?**

The referencing could do relative path logic or not. That shoudl likely be configurable. In the base case for CommonJS the `__filename` value would be the output file `__filename` value which seemed to work in some scenarios.

One could just define CJS free `__filename` references as `fileURLToPath(import.meta.url)` in the Webpack bundle.

**Are you willing to work on this yourself?**
no

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**

feature

closes #14072 

**Did you add tests for your changes?**

yes

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

new options node.__filename=node-module / node.__dirname=node-module

generate `__filename` and `__dirname` for common js modules when output.module=true to
`fileURLToPath(import.meta.url)` and `fileURLToPath(import.meta.url + "/..")` respectively

### CODE REVIEW

File: test/configCases/output-module/node-globals/index.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { dir, file } from './cjs/file.js'
+
+it("should generate correct __dirname", () => {
+	expect(dir.slice(dir.lastIndexOf("/", dir.length - 2))).toBe("/node-globals/");
```
Comment (sokra) [inline test/configCases/output-module/node-globals/index.js:4]:
```suggestion
	expect(dir.slice(dir.lastIndexOf("/", dir.length - 2))).toMatch(/[\\/]node-globals[\\/]/);
```

**Review by sokra — CHANGES_REQUESTED** [review]

**Review by sokra — COMMENTED** [review]
Found these old review comments... not sure if they are still relevant

File: lib/dependencies/NodeUrlDependency.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		/** @type {InitFragment<ChunkRenderContext>[]} */
+		let chunkInitFragments = data.get("chunkInitFragments");
+
+		if (!chunkInitFragments) {
+			chunkInitFragments = [];
+			data.set("chunkInitFragments", chunkInitFragments);
+		}
+
+		chunkInitFragments.push(
+			new InitFragment(
```
[the remark this answers is not in the document]
Comment (vankop) [reply lib/dependencies/NodeUrlDependency.js]: did not fix this yet

[the remark this answers is not in the document]
Comment (vankop) [reply lib/dependencies/NodeUrlDependency.js]: fixed

File: test/configCases/output-module/node-globals/index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { dir, file } from './cjs/file.js'
+
+it("should generate correct __dirname", () => {
+	const match = dir.match(/[\\/][^\\/]+[\\/]$/);
```
[the remark this answers is not in the document]
Comment (vankop) [reply test/configCases/output-module/node-globals/index.js]: done

**Review by alexander-akait — APPROVED** [review]
@vankop can you rebase, I think we can merge it and I am fine with it, thank you

**Comment by iamyamakin** [comment]
@sokra do you mind if i create PR with a merge commit? I see a few conflicts between this branch and main:

- lib/NodeStuffPlugin.js;
- lib/config/defaults.js;
- schemas/WebpackOptions.check.js.

**Comment by snitin315** [comment]
I'll rebase this PR, no worries.

**Comment by alexander-akait** [comment]
Yeah, feel free to send again, I am fine with the such improve

**Pull request merged**